R/formula.R

```diff
@@ -0,0 +1,283 @@
+#' Specifically, it is not possible to use [stats::offset] in a formula. Analyses that allow including [stats::offset] in the model have specialized argument for that purpose.
+#' It is also not possible to use variable transformations in a formula. Thus, instead of transforming variables using formulas, transform the variables before entering them in the analysis.
+#'
+#' For specification of the random effects, [lme4::lme4-package] syntax is used. There is a difference in how JASP parses whether or not are random effects correlated:
+#' Under each random grouping factor, if some but not all terms are correlated, the output `correlated` is still set to \code{TRUE}. The "correlations" attribute contains the full correlation structure.
+#' @example inst/examples/ex-formula.R
+#' @rdname jaspFormula
+#' @export
+jaspFormula <- function(formula, data) {
+  formulaCheckRequirements(formula, data)
+
+  result <- list(
+    formula = formula,
+    lhs = formulaGetLhs(formula),
+    rhs = formulaGetRhs(formula)
+  )
+
+  class(result) <- c("jaspFormula")
+  return(result)
+}
+
+#
+#' @rdname jaspFormula
+#' @export
+createJaspFormula <- function(..., response=NULL, data) {
+  if(!is.data.frame(data)) {
+    stop("`data` must be a data frame.")
+  }
+
+  if(!is.null(response) && !is.character(response)) {
+    stop("`response` must be a character.")
+  }
+
+  dots <- list(...)
+  if(!is.jaspRhs(dots))
+    stop("All terms on the rhs must be specified as a `jaspFormulaRhs` object.")
+
+  if(length(response) > 1) {
+    response <- paste0("cbind(", paste0(response, collapse = ","), ")")
+  }
+
+
+  rhs <- vapply(dots, createJaspFormulaRhs, character(1), data = data)
+
+  formula <- reformulate(rhs, response)
+  return(jaspFormula(formula, data))
+}
+
+
+#
+#' @rdname jaspFormula
+#' @export
+jaspFormulaRhs <- function(terms = NULL, group = NULL, intercept = TRUE, correlated = TRUE) {
+  result <- list(terms = terms, group = group, intercept = intercept, correlated = correlated)
+  class(result) <- "jaspFormulaRhs"
+  return(result)
+}
+
+#' @rdname jaspFormula
+#' @export
+as.character.jaspFormula <- function(x) {
+  deparse(x[["formula"]])
+}
+
+createJaspFormulaRhs <- function(rhs, data) {
+  allVarNames <- colnames(data)
+
+  result <- paste(rhs[["terms"]], collapse = "+")
+  if(!rhs[["intercept"]]) {
+    result <- paste("0", result, sep = "+")
+  } else if(is.null(rhs[["terms"]])) {
+    result <- "1"
+  }
+  if(!is.null(rhs[["group"]])) {
+    sep <- if(rhs[["correlated"]]) "|" else "||"
+    result <- sprintf("(%s %s %s)", result, sep, rhs[["group"]])
+  }
+  return(result)
+}
+
+is.jaspRhs <- function(x) {
+  if(inherits(x, "list")) {
+    all(vapply(x, is.jaspRhs, logical(1)))
+  } else {
+    inherits(x, "jaspFormulaRhs")
+  }
+}
+
+formulaCheckRequirements <- function(formula, data) {
+  if (!inherits(formula, "formula")) {
+    stop("`formula` argument must be object of class `formula`.")
```

Comment:
> please add this to all `stop`s and `warning`s, otherwise they show up in weblate for translation, which I think is not necessary.
In order to be translated in WebLate, you need also to use gettext

Reply:
I agree with Don, since I find that some unnecessary strings appeared in [WebLate](https://hosted.weblate.org/translate/jasp/jaspbase/en/?checksum=783f7bd84a6b87b0&sort_by=-priority,position) from [here](https://github.com/jasp-stats/jaspBase/blob/5c53dad380c1536d3bd5dc5cf647bcde3a52ba62/R/moduleInstall.R#L241) but without a gettext call.

Reply:
@boutinb unless we tweaked `xgettext` to not record these for `stop` and `warning`, this is not the case, see https://stat.ethz.ch/R-manual/R-devel/library/base/html/stop.html:

![image](https://user-images.githubusercontent.com/21319932/198546717-8828c803-4981-44b3-9ea5-e1a2a770dcf1.png)

By default, R lists all messages in `stop` and `warning` to be translated. I also noticed this in jaspGraphs when a bunch of strings were translated that didn't need to be translated (e.g., error messages intended for developers).

Reply:
@shun2wang that is indeed an example where we should have used `domain = NA` to not translate the string, or `gettextf` so the string can be translated properly.

Reply:
thanks, done